src/decaf.cpp

```diff
@@ -159,6 +173,84 @@ namespace decaf {
     return ExecutionResult::Continue;
   }
 
+  ExecutionResult Interpreter::visitBranchInst(llvm::BranchInst& inst) {
+    auto jump_to = [&](llvm::BasicBlock* target) {
+      auto& frame = ctx->stack_top();
+      frame.prev_block = frame.current_block;
+      frame.current_block = target;
+    };
+    
+    if (!inst.isConditional()) {
+      jump_to(inst.getSuccessor(0));
+      return ExecutionResult::Continue;
+    }
+
+    auto& frame = ctx->stack_top();
+    auto cond = normalize_to_bool(frame.lookup(inst.getCondition(), *z3));
+
+    auto is_t = ctx->check(cond);
+    auto is_f = ctx->check(!cond);
+
+    // Note: For the purposes of branching we consider unknown to be
+    //       equivalent to sat. Maybe future branches will bring the
+    //       equation back to being solvable.
+    if (is_t != z3::unsat && is_f != z3::unsat) {
+      auto fork = ctx->fork();
+
+      // In cases where both conditions are possible we follow the
+      // false path. This should be enough to get us out of most loops
+      // and actually exploring the rest of the program.
+      fork.add(cond);
+      ctx->add(!cond);
+
+      queue->add_context(std::move(fork));
+      return ExecutionResult::Continue;
+    } else if (is_t != z3::unsat) {
+      ctx->add(cond);
+      return ExecutionResult::Continue;
+    } else if (is_f != z3::unsat) {
+      ctx->add(!cond);
+      return ExecutionResult::Continue;
+    } else {
+      return ExecutionResult::Stop;
+    }
+  }
+
+  ExecutionResult Interpreter::visitReturnInst(llvm::ReturnInst& inst) {
+    auto& frame = ctx->stack_top();
+
+    std::optional<z3::expr> result;
+    if (inst.getNumOperands() != 0)
+      result = frame.lookup(inst.getOperand(0), *z3);
+
+    ctx->stack.pop_back();
+
+    // Reached end of function, nothing left to do.
+    if (ctx->stack.empty())
+      return ExecutionResult::Stop;
+
+    if (result.has_value()) {
+      auto& parent = ctx->stack_top();
+      auto& caller = *std::prev(parent.current);
```

Comment:
Why is the previous frame of the parent the caller? Didn't the current frame get popped on line 226, making `stack_top` the caller?

Reply:
This is getting the previous instruction in the parent frame, not the previous frame. Basically, we've already executed the call instruction in the parent frame so we need to go back one instruction to get to it.

Reply:
Ok, I see. The return value of the call instruction is the call instruction